### ISSUE TITLE
perf(quotient): baseline the quotient stage vs native pil2 (calibrated proxy ~2×, needs real constraints to confirm)

### ISSUE THREAD

**Issue by erkhesbold2005-blip** [fractalyze/zisk-zorch#66]
## Goal

Establish a fair, calibrated baseline for the **quotient stage** (`quotient_from_constraints` — `constraint_eval` fold + zerofier divide) vs the native pil2-stark CUDA prover, and close the gap if one is confirmed. A calibrated proxy currently reads ~2× slower than pil2, but the proxy is bandwidth-pessimistic — the **true** number requires running the real re-authored Main constraints, which this issue tracks.

## Why

Baselining the per-stage prover on an RTX 5090 leaves quotient as the last stage without a trustworthy number. It matters most: quotient is ~55% of per-air compute. The other stages: extend 0.58× (win), commit 1.04× (parity), logup grand-sum 1.30× (#64), FRI 2.52× (#65). Quotient is the second stage where zorch may trail.

## Measured (RTX 5090, N_ext = 2²³)

pil2's Main quotient is a **bytecode interpreter** (`calculateExpressions`, a per-row VM stepping an `expressions.bin` op stream; `expressions_gpu.cu`). Its runnable density proxy `MAIN_EXPR_PATTERN` (tuned to the real ~135 ms Q-kernel) does **~7000 base-field FMAs + ~900 Fp3 mul-adds per row**, register-resident, over 2²³.

- **pil2 `MAIN_EXPR` = 134 ms** (measured fresh; brackets constraint-fold + embedded ZHInv zerofier multiply, **excludes** the INTT-back and Merkle).
- **Our proxy, calibrated to that density** (`bench_inner_proof --stages=quotient --n_constraints=900 --degree=9` → ~7200 base mul + 900 Fp3/row): per-row saturated (34.3 ns @2²⁰ → 32.8 ns @2²¹), extrapolating to **~270 ms @2²³ ≈ 2.0× slower**. Our zerofier `divide` alone is 0.32 ms (matches pil2's embedded-zerofier bracket).
- Our old default proxy (64 constraints × degree-3) was **1/55th** of Main's op density — the historical "45 ms" was drastically undersized, not a real win.

## Caveat — why this is not yet a clean baseline

Our synthetic proxy is **900 independent degree-9 column products**, so it **materializes a large row-scaled intermediate** (~9.8 GB of scratch beyond the 1 GB trace at 2²¹) → HBM-bandwidth-bound. pil2's proxy and the real interpreter keep operands **register-resident** (16 lanes, FMA chains) → no materialization. So the ~270 ms is a **pessimistic upper bound on structure**; the real zisk-zorch quotient — reauthor.py's actual Main constraints, a shared-subexpression DAG that fuses through `zorch.constraint_eval` — would very likely be faster. **The direction (slower vs competitive) is unresolved until the real constraints are measured.**

Architecturally this cuts in our favor: pil2 is an **interpreter**, ours is a **compiled fused XLA kernel**. A register-optimal compiled quotient *should* be competitive-to-faster than a per-row VM; our current bandwidth-heavy proxy just doesn't demonstrate it.

## Plan

1. **Get the true number**: run reauthor.py's actual Main-air constraints through `constraint_eval` at 2²³ and compare to pil2's 134 ms. This needs the rw/bazel setup (`load_zisk_chips`, `rw_constraints`) — not runnable in the bare gpu-venv (same "no real Main starkinfo vendored" wall as FRI #65). **This is the gating step** — the proxy is only a stand-in.
2. **If confirmed slower**, investigate whether `constraint_eval` materializes intermediates instead of keeping the constraint DAG register/fusion-resident (the proxy's failure mode). Levers likely in **zorch** (`constraint_eval` primitive) — cross-repo, cf. #65 → zorch and #58 → xla. The cubic SoA layout (xla_fork#258) would also help the Fp3 portion.

## End criteria

- [ ] Real Main-air quotient (reauthor `constraint_eval`, 2²³) measured head-to-head vs pil2's 134 ms — a trustworthy ratio, not the proxy.
- [ ] If slower, the dominant cost (constraint-DAG materialization vs register-resident eval) is identified and a lever proposed.
- [ ] `quotient_test.py` / stage-2 byte-match preserved by any change.

## Repro / harness

- Ours (proxy): `bench_inner_proof --stages=quotient --n_bits=<=20 --n_constraints=900 --degree=9` (gpu-venv). ⚠️ 900 constraints × 2²²⁺ rows OOMs (~44 GB @2²³) — measure ≤2²¹ and extrapolate (per-row is saturated/constant).
- pil2: `/tmp/claude-1006/main_bench --benchmark_filter='MAIN_EXPR'` = 134 ms (density mimic at Main's 38/24/6-dim-3 columns, 2²³).
- Note: unrelated to the LogUp grand-sum work (#64, prime-ir #398).

**Comment by erkhesbold2005-blip** [comment]
## The ~270 ms / OOM figures do not reproduce on main's shipped pins

Re-measured 2026-07-15 on the same RTX 5090 (driver 595.71.05), against
`origin/main`'s pins (`zorch==0.1.0.dev20260714055619`,
`jax==jaxlib==jax-cuda12-pjrt==0.10.0.dev20260714051543`, `zk-dtypes==0.0.10`) —
no local override, no swapped plugin. Both of this issue's load-bearing claims
fail:

| this issue | measured |
|---|---|
| "900 constraints × 2^22+ rows OOMs (~44 GB @2^23)" | **runs at 2^23; peak 5.13 GiB** |
| "~270 ms @2^23" (extrapolated from ≤2^21) | **77.0 ms** |
| "~2.0× slower" vs `MAIN_EXPR` | **0.58×** (native re-measured at 133 ms) |

### Why the extrapolation missed by ~3.5×

The proxy **changes compilation regime** between 2^21 and 2^22 rows, so the
per-row cost this issue extrapolated from is not the per-row cost at the real
size:

| rows | latency | ns/row | peak | regime |
|---|---|---|---|---|
| 2^20 | 35.22 ms | 33.6 | 4.95 GiB | materialized |
| 2^21 | 68.66 ms | 32.7 | 9.91 GiB | materialized |
| 2^22 | 37.86 ms | **9.0** | 2.56 GiB | **fused** |
| 2^23 | 77.11 ms | **9.2** | 5.13 GiB | **fused** |

Note 2^21 → 2^22 goes **down** (68.66 → 37.86 ms) while the size doubles: above
2^21 XLA stops materializing the intermediates and fuses them, and the ~9.8 GB
scratch this issue identifies as the HBM bottleneck disappears (9.91 → 2.56 GiB).
Extrapolating 32.8 ns/row from the materialized regime gives ~270 ms; the fused
2^23 actually runs at 9.2 ns/row.

The work is real, not elided — latency scales with the constraint count at 2^23:

| `--n_constraints` | 225 | 450 | 900 |
|---|---|---|---|
| latency | 33.22 ms | 47.59 ms | 76.96 ms |

(≈19 ms fixed + ~0.064 ms/constraint.) Reproduces across fresh processes.

### Repro

```sh
# ours (per row of the table; --n_bits+1 = rows)
CUDA_VISIBLE_DEVICES=0 XLA_PYTHON_CLIENT_PREALLOCATE=false PYTHONPATH=<zisk-zorch> \
  python -m zisk_zorch.bench_inner_proof --stages=quotient \
    --n_bits=22 --n_cols=38 --blowup_bits=1 --n_constraints=900 --degree=9 \
    --phase runtime -o q.json

# native
cd /tmp/claude-1006 && LD_LIBRARY_PATH=/tmp/claude-1006/gmp-prefix/lib CUDA_VISIBLE_DEVICES=0 \
  ./main_bench --benchmark_filter='MAIN_EXPR_PATTERN' --benchmark_min_time=5x \
    --benchmark_repetitions=3 --benchmark_report_aggregates_only=true
```

⚠️ A venv built from `requirements.in` alone is **CPU-only** (the pins name
`jax-cuda12-plugin` but none of the `nvidia-*-cu12` libs it loads) and will
silently benchmark the CPU. Install `jax-cuda12-plugin[with-cuda]` at the same
pin and assert `jax.devices()` shows `CudaDevice`.

### What this does and does not change

- **"quotient is ~2× slower" is unsupported.** Measured, our proxy beats the
  native proxy. The `~270 ms` figure should not be re-quoted — same failure mode
  as the "45 ms" number this issue already retracted, one regime up.
- **This issue's core ask stands.** Both sides are still proxies: ours is 900
  independent degree-9 products, `MAIN_EXPR_PATTERN` is a density mimic tuned to
  the real ~135 ms kernel, not the bytecode interpreter. **0.58× is not a
  baseline** — settling the direction still needs the real Main constraints
  through `constraint_eval`, exactly as scoped here.
- The HBM-bound diagnosis is now at least size-dependent: at the real size the
  proxy fuses.

Recorded in `docs/zisk-baseline.md` (#60).

**Comment by erkhesbold2005-blip** [comment]
## Retracting this comment's numbers — the bench proxy was CSE-collapsed

Everything measured above is invalid. `bench_inner_proof._make_eval_fn` did not
build the density it reported, so every latency and footprint figure in this
comment describes a workload ~24× smaller than `--n_constraints=900` asked for.

### The bug

The old picker indexed columns as

```python
c = t[:, k % n_cols]
for d in range(1, degree):
    c = c * t[:, (k * degree + d) % n_cols]
```

Both indices depend only on `k mod n_cols`, so the column tuple repeats every
`n_cols` constraints. At `--n_cols=38 --n_constraints=900` that is **38 distinct
degree-9 products, CSE-folded, stacked 900 wide** — not 900 products. Fixed in
#77 with a seeded per-constraint draw (`np.random.default_rng(k).choice`).

Latency still scaled with `--n_constraints` (the 225/450/900 → 33/48/77 ms table
above), which I read as proof the work was real. That inference was wrong: the
output stack is `(rows, n_constraints)` and materializes at full width no matter
how few distinct products feed it. The table measured output bandwidth, not
constraint arithmetic — "≈19 ms fixed + 0.064 ms/constraint" is the cost of
writing columns, not of evaluating them.

### What is retracted

| claimed above | status |
|---|---|
| "**77.0 ms** @2^23" | retracted — CSE-collapsed |
| "**0.58×** vs `MAIN_EXPR`" | retracted — same source |
| "runs at 2^23; peak **5.13 GiB**" | retracted — the corrected proxy OOMs at **2^22** |
| regime-change table (33 → 9.2 ns/row) | retracted — there is no fusion; see below |

### There was never a fusion regime change

I claimed XLA stops materializing above 2^21. Re-measured on the fixed bench
(same RTX 5090, main's pins, `c440b80`), the corrected proxy materializes ~900
full-height base-field intermediates and never fuses. The `900 × rows × 8 B`
model predicts it closely:

| rows | predicted | measured |
|---|---|---|
| 2^20 | 7.03 GiB | 47.4 ms |
| 2^21 | 14.06 GiB | **13.65 GiB peak**, 92.5 ms |
| 2^22 | 28.1 GiB | **OOM** — a single 21.97 GiB allocation fails |
| 2^23 | 56.2 GiB | unreachable |

Flat ~44 ns/row, ceiling at 2^22 against ~23.5 GiB usable (frx caps at ~75% of
31.8 GiB).

The old "fused" regime is fully explained by the collapse: 38 intermediates at
2^23 predict ≈2.4 GiB, and I measured 2.56 GiB. That was 38 products fitting
comfortably, not 900 products fusing.

### This withdraws my retraction of "~270 ms"

The fusion story was the whole basis for calling this issue's "~270 ms" an
extrapolation artifact. That basis is gone, and the correction cuts the other
way: extrapolating the corrected 44 ns/row to 2^23 gives **~370 ms**, the same
order as "~270 ms", not 3.5× off.

I am not quoting ~370 ms either. It extrapolates two doublings past a hard OOM
boundary — the identical move that produced both bad numbers in this thread. The
honest position is that **the quotient proxy has no number at Main's size, in
either direction**, and that the "~270 ms" retraction should not have been
posted.

### What stands, and is now better supported

- **This issue's core ask, unchanged and now the only path.** Both sides are
  proxies; `MAIN_EXPR_PATTERN` is 2^23-only with no sweep, so no same-size pair
  exists and no ratio is quotable at all. Settling the direction needs the real
  Main constraints through `constraint_eval`, which reuse columns instead of
  forming 900 independent products.
- **The HBM-bound diagnosis, sharpened.** I said above it was "at least
  size-dependent: at the real size the proxy fuses." It is not — it is
  size-independent and worse than this issue framed it. We materialize where
  pil2's bytecode interpreter keeps operands register-resident, which is the same
  shape as #69 and is why the proxy cannot even reach the size the native runs
  at.
- The venv/GPU warning above.

Repro is unchanged from this comment, against `c440b80` (`--n_bits`+1 = rows).
Recorded in `docs/development.md` (#60 / #77).
